### PR TITLE
Add option "-fsycl-allow-device-image-dependencies" 

### PR DESCRIPTION
**Pull request by LU-JOHN**
Add option "-fsycl-allow-device-image-dependencies".  This option replaces "-fsycl-allow-device-dependencies" which is deprecated.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -7025,8 +7025,14 @@ def fsycl_remove_unused_external_funcs : Flag<["-"], "fsycl-remove-unused-extern
...
 def fsycl_allow_device_dependencies : Flag<["-"], "fsycl-allow-device-dependencies">,
-  HelpText<"Allow dependencies between device code images">;
+  HelpText<"Deprecated: please use -fsycl-allow-device-image-dependencies instead.">,
+  Flags<[Deprecated]>;
 def fno_sycl_allow_device_dependencies : Flag<["-"], "fno-sycl-allow-device-dependencies">,
+  HelpText<"Deprecated: please use -fno-sycl-allow-device-image-dependencies instead.">,
+  Flags<[Deprecated]>;
+def fsycl_allow_device_image_dependencies : Flag<["-"], "fsycl-allow-device-image-dependencies">,
+  HelpText<"Allow dependencies between device code images">;
+def fno_sycl_allow_device_image_dependencies : Flag<["-"], "fno-sycl-allow-device-image-dependencies">,
```
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:7035]: This looks like a good opportunity to use the `BoolOptionWithoutMarshalling` macro.  You can use `sycl_add_default_spec_consts_image` as a reference.

File: clang/test/Driver/sycl-offload-old-model.cpp [hunk 2]
```diff
@@ -174,16 +174,16 @@
...
+// RUN: %clang -### -target x86_64-unknown-linux-gnu -fsycl --no-offload-new-driver -fsycl-targets=spir64_gen -fsycl-allow-device-image-dependencies %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_ADID_PASS %s
+// RUN: %clang -### -target x86_64-unknown-linux-gnu -fsycl --no-offload-new-driver -fsycl-targets=spir64_gen -fno-sycl-allow-device-image-dependencies %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_ADID_NO_PASS %s
+// RUN: %clang_cl -### -fsycl --no-offload-new-driver -fsycl-targets=spir64_fpga -fsycl-allow-device-image-dependencies %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_ADID_PASS %s
+// RUN: %clang_cl -### -fsycl --no-offload-new-driver -fsycl-targets=spir64_gen -fsycl-allow-device-image-dependencies %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_ADID_PASS %s
+// RUN: %clang_cl -### -fsycl --no-offload-new-driver -fsycl-targets=spir64_gen -fno-sycl-allow-device-image-dependencies %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_ADID_NO_PASS %s
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload-old-model.cpp:186]: Can you also add a test to `sycl-deprecated.cpp` for the old option?

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (LU-JOHN) [reply clang/include/clang/Driver/Options.td]: Updated to use `BoolOptionWithoutMarshalling`.

File: clang/test/Driver/sycl-offload-old-model.cpp (hunk 2 above)
Comment (LU-JOHN) [reply clang/test/Driver/sycl-offload-old-model.cpp:186]: Added.

**Review by mdtoguchi — APPROVED** [review]
OK for driver

**Review by aelovikov-intel — APPROVED** [review]

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -10791,10 +10791,17 @@ static void addArgs(ArgStringList &DstArgs, const llvm::opt::ArgList &Alloc,
   }
 }
 
-static bool allowDeviceDependencies(const llvm::opt::ArgList &TCArgs) {
+static bool allowDeviceImageDependencies(const llvm::opt::ArgList &TCArgs) {
+  // deprecated
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:10795]: Do we really need this?

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks

**Comment by aelovikov-intel** [comment]
```
Failed Tests (3):
SYCL :: DeviceDependencies/free_function_kernels.cpp
SYCL :: LLVMIntrinsicLowering/bitreverse.cpp
SYCL :: LLVMIntrinsicLowering/sub_byte_bitreverse.cpp
```

seem to be failing in both pre/post-commit after this (https://github.com/intel/llvm/actions/runs/11018616463)

@LU-JOHN , @asudarsa

**Comment by AlexeySachkov** [comment]
> ```
> Failed Tests (3):
> SYCL :: DeviceDependencies/free_function_kernels.cpp
> SYCL :: LLVMIntrinsicLowering/bitreverse.cpp
> SYCL :: LLVMIntrinsicLowering/sub_byte_bitreverse.cpp
> ```
> 
> seem to be failing in both pre/post-commit after this (https://github.com/intel/llvm/actions/runs/11018616463)
> 
> @LU-JOHN , @asudarsa

#15511

**Pull request merged**